src/session.rs

```diff
@@ -0,0 +1,137 @@
+use std::collections::HashMap;
+use std::convert::TryFrom;
+use std::fs;
+use std::path::PathBuf;
+
+use anyhow::Result;
+use cookie::Cookie;
+use reqwest::header::{HeaderMap, HeaderValue, AUTHORIZATION, COOKIE, SET_COOKIE};
+use serde::{Deserialize, Serialize};
+
+#[derive(Debug, Serialize, Deserialize)]
+struct Content {
+    #[serde(rename = "__meta__")]
+    meta: HashMap<String, String>,
+    // TODO: Replace String with Cookie
+    // (need to use Serde's serialize_with and deserialize_with)
+    cookies: HashMap<String, String>,
+    headers: HashMap<String, String>,
+}
+
+impl Default for Content {
+    fn default() -> Self {
+        let mut meta = HashMap::new();
+        meta.insert("about".into(), "xh session file".into());
+        meta.insert("xh".into(), env!("CARGO_PKG_VERSION").into());
+        Content {
+            meta,
+            cookies: HashMap::new(),
+            headers: HashMap::new(),
+        }
+    }
+}
+
+pub struct Session {
+    pub path: PathBuf,
+    pub read_only: bool,
+    content: Content,
+}
+
+fn ensure_file_exists(path: PathBuf) -> Result<PathBuf> {
+    if path.exists() {
+        Ok(path)
+    } else {
+        fs::create_dir_all(path.parent().unwrap())?;
+        fs::File::create(&path)?;
+        Ok(path)
+    }
+}
+
+impl Session {
+    pub fn load_session(host: &str, name_or_path: &str, read_only: bool) -> Result<Self> {
+        let path = if name_or_path.contains(std::path::is_separator) {
+            ensure_file_exists(PathBuf::from(name_or_path))?
+        } else {
+            let mut path = dirs::config_dir()
+                .unwrap()
+                .join::<PathBuf>(["xh", "sessions", host, name_or_path].iter().collect());
+            path.set_extension("json");
```

Comment:
This turns `foo.bar.baz` into `foo.bar.json`, where HTTPie turns it into `foo.bar.baz.json`. So a dumb append would be better.